File: frontend/src/services/adminJob.service.js
import adminJobApi from "../api/adminJob.api";

import { compareDateDesc } from "../utils/dateTime.utils";
const getValue = (...values) => {
  return values.find(
    (value) => value !== undefined && value !== null && value !== ""
  );
};

const toNumber = (value, fallback = 0) => {
  const number = Number(value);
  return Number.isNaN(number) ? fallback : number;
};

const unwrapData = (response) => {
  const data = response?.data;

  if (!data) return null;

  if (data?.data?.job) return data.data.job;
  if (data?.data?.item) return data.data.item;
  if (data?.data?.result) return data.data.result;
  if (data?.data !== undefined) return data.data;

  if (data?.job) return data.job;
  if (data?.item) return data.item;
  if (data?.result) return data.result;

  return data;
};

const unwrapListData = (response) => {
  const data = response?.data;

  if (Array.isArray(data)) return data;

  if (Array.isArray(data?.data)) return data.data;
  if (Array.isArray(data?.items)) return data.items;
  if (Array.isArray(data?.result)) return data.result;
  if (Array.isArray(data?.jobs)) return data.jobs;
  if (Array.isArray(data?.proposals)) return data.proposals;

  if (Array.isArray(data?.data?.items)) return data.data.items;
  if (Array.isArray(data?.data?.result)) return data.data.result;
  if (Array.isArray(data?.data?.jobs)) return data.data.jobs;
  if (Array.isArray(data?.data?.proposals)) return data.data.proposals;

  return [];
};

const normalizeJob = (job) => {
  if (!job) return null;

  const jobId = getValue(
    job.jobPostingId,
    job.JobPostingId,
    job.jobId,
    job.JobId,
    job.id,
    job.Id,
    job.jobID,
    job.JobID
  );

  return {
    jobId,
    id: jobId,

    title: getValue(
      job.title,
      job.Title,
      job.jobTitle,
      job.JobTitle,
      "Untitled Job"
    ),

    description: getValue(
      job.description,
      job.Description,
      job.jobDescription,
      job.JobDescription,
      job.summary,
      job.Summary,
      ""
    ),

    status: String(getValue(job.status, job.Status, "OPEN")).toUpperCase(),

    clientId: getValue(job.clientId, job.ClientId, null),

    clientProfileId: getValue(job.clientProfileId, job.ClientProfileId, null),
    clientUserId: getValue(job.clientUserId, job.ClientUserId, null),

    clientName: getValue(
      job.clientName,
      job.ClientName,
      job.client?.fullName,
      job.Client?.FullName,
      job.client?.name,
      job.Client?.Name,
      "Client"
    ),

    clientEmail: getValue(job.clientEmail, job.ClientEmail, ""),
    clientType: getValue(job.clientType, job.ClientType, ""),

    category: getValue(
      job.category,
      job.Category,
      job.expertCategory,
      job.ExpertCategory,
      ""
    ),

    level: getValue(
      job.level,
      job.Level,
      job.experienceLevel,
      job.ExperienceLevel,
      ""
    ),

    projectType: getValue(job.projectType, job.ProjectType, ""),

    complexity: getValue(job.complexity, job.Complexity, ""),

    expectedDeliverables: getValue(
      job.expectedDeliverables,
      job.ExpectedDeliverables,
      ""
    ),

    budgetMin: toNumber(
      getValue(job.budgetMin, job.BudgetMin, job.minBudget, job.MinBudget, 0)
    ),

    budgetMax: toNumber(
      getValue(job.budgetMax, job.BudgetMax, job.maxBudget, job.MaxBudget, 0)
    ),

    budget: toNumber(getValue(job.budget, job.Budget, job.amount, job.Amount, 0)),

    deadline: getValue(job.deadline, job.Deadline, ""),
    publishedAt: getValue(job.publishedAt, job.PublishedAt, ""),

    isAiAssisted: Boolean(getValue(job.isAiAssisted, job.IsAiAssisted, false)),
    aiGeneratedDescription: getValue(
      job.aiGeneratedDescription,
      job.AiGeneratedDescription,
      ""
    ),
    postingChargeType: getValue(
      job.postingChargeType,
      job.PostingChargeType,
      ""
    ),

    proposalCount: toNumber(
      getValue(
        job.proposalCount,
        job.ProposalCount,
        job.totalProposals,
        job.TotalProposals,
        0
      )
    ),

    submittedProposalCount: toNumber(
      getValue(job.submittedProposalCount, job.SubmittedProposalCount, 0)
    ),
    acceptedProposalCount: toNumber(
      getValue(job.acceptedProposalCount, job.AcceptedProposalCount, 0)
    ),
    rejectedProposalCount: toNumber(
      getValue(job.rejectedProposalCount, job.RejectedProposalCount, 0)
    ),
    withdrawnProposalCount: toNumber(
      getValue(job.withdrawnProposalCount, job.WithdrawnProposalCount, 0)
    ),

    skills: getValue(job.skills, job.Skills, []),

    createdAt: getValue(job.createdAt, job.CreatedAt, job.postedAt, job.PostedAt, ""),
    updatedAt: getValue(job.updatedAt, job.UpdatedAt, ""),

    raw: job,
  };
};

const normalizeProposal = (proposal) => {
  if (!proposal) return null;

  const proposalId = getValue(
    proposal.proposalId,
    proposal.ProposalId,
    proposal.id,
    proposal.Id
  );

  return {
    proposalId,
    id: proposalId,

    jobId: getValue(proposal.jobId, proposal.JobId, null),
    jobTitle: getValue(proposal.jobTitle, proposal.JobTitle, ""),

    clientProfileId: getValue(
      proposal.clientProfileId,
      proposal.ClientProfileId,
      null
    ),
    clientUserId: getValue(proposal.clientUserId, proposal.ClientUserId, null),
    clientName: getValue(proposal.clientName, proposal.ClientName, "Client"),
    clientAvatarUrl: getValue(
      proposal.clientAvatarUrl,
      proposal.ClientAvatarUrl,
      ""
    ),

    expertProfileId: getValue(
      proposal.expertProfileId,
      proposal.ExpertProfileId,
      null
    ),
    expertUserId: getValue(proposal.expertUserId, proposal.ExpertUserId, null),

    expertName: getValue(
      proposal.expertName,
      proposal.ExpertName,
      proposal.expert?.fullName,
      proposal.Expert?.FullName,
      "Expert"
    ),

    coverLetter: getValue(
      proposal.coverLetter,
      proposal.CoverLetter,
      proposal.description,
      proposal.Description,
      ""
    ),

    expertAvatarUrl: getValue(
      proposal.expertAvatarUrl,
      proposal.ExpertAvatarUrl,
      ""
    ),

    proposedBudget: toNumber(
      getValue(
        proposal.proposedPrice,
        proposal.ProposedPrice,
        proposal.proposedBudget,
        proposal.ProposedBudget,
        proposal.amount,
        proposal.Amount,
        0
      )
    ),

    proposedPrice: toNumber(
      getValue(
        proposal.proposedPrice,
        proposal.ProposedPrice,
        proposal.proposedBudget,
        proposal.ProposedBudget,
        proposal.amount,
        proposal.Amount,
        0
      )
    ),

    proposedTimelineDays: toNumber(
      getValue(
        proposal.proposedTimelineDays,
        proposal.ProposedTimelineDays,
        proposal.timelineDays,
        proposal.TimelineDays,
        0
      )
    ),

    expectedOutputs: getValue(
      proposal.expectedOutputs,
      proposal.ExpectedOutputs,
      ""
    ),
    workingApproach: getValue(
      proposal.workingApproach,
      proposal.WorkingApproach,
      ""
    ),
    preliminaryMilestonePlan: getValue(
      proposal.preliminaryMilestonePlan,
      proposal.PreliminaryMilestonePlan,
      ""
    ),
    milestones: getValue(proposal.milestones, proposal.Milestones, []),

    status: String(getValue(proposal.status, proposal.Status, "PENDING")).toUpperCase(),
    statusReason: getValue(proposal.statusReason, proposal.StatusReason, ""),
    contractId: getValue(proposal.contractId, proposal.ContractId, null),
    clientMissSignCount: toNumber(
      getValue(proposal.clientMissSignCount, proposal.ClientMissSignCount, 0)
    ),
    expertMissSignCount: toNumber(
      getValue(proposal.expertMissSignCount, proposal.ExpertMissSignCount, 0)
    ),
    latestVersionNumber: toNumber(
      getValue(proposal.latestVersionNumber, proposal.LatestVersionNumber, 1)
    ),
    totalVersions: toNumber(
      getValue(proposal.totalVersions, proposal.TotalVersions, 1)
    ),
    lastResubmittedAt: getValue(
      proposal.lastResubmittedAt,
      proposal.LastResubmittedAt,
      null
    ),
    latestVersion: getValue(proposal.latestVersion, proposal.LatestVersion, null),

    createdAt: getValue(proposal.createdAt, proposal.CreatedAt, ""),

    raw: proposal,
  };
};

const adminJobService = {
  async getAllJobs(params = {}) {
    const response = await adminJobApi.getAllJobs(params);
    return unwrapListData(response)
      .map(normalizeJob)
      .filter(Boolean)
      .sort((a, b) =>
        compareDateDesc(
          a.updatedAt || a.createdAt,
          b.updatedAt || b.createdAt
        )
      );
  },

  async getJobById(jobId) {
    if (!jobId || jobId === "undefined" || jobId === "null") {
      throw new Error("Invalid job id.");
    }

    const response = await adminJobApi.getJobById(jobId);
    return normalizeJob(unwrapData(response));
  },

  async getJobProposals(jobId) {
    if (!jobId || jobId === "undefined" || jobId === "null") {
      throw new Error("Invalid job id.");
    }

    const response = await adminJobApi.getJobProposals(jobId);
    return unwrapListData(response)
      .map(normalizeProposal)
      .filter(Boolean)
      .sort((a, b) => compareDateDesc(a.createdAt, b.createdAt));
  },

  async cancelJob(jobId, reason = "") {
    if (!jobId || jobId === "undefined" || jobId === "null") {
      throw new Error("Invalid job id.");
    }

    const payload = {
      reason: String(reason || "").trim(),
    };

    const response = await adminJobApi.cancelJob(jobId, payload);
    return normalizeJob(unwrapData(response));
  },

  normalizeJob,
  normalizeProposal,
};

export default adminJobService;

File: backend/AITasker.Infrastructure/Services/AdminJobService.cs
using AITasker.Application.DTOs.Requests;
using AITasker.Application.DTOs.Responses;
using AITasker.Application.Interfaces;
using AITasker.Domain.Constants;
using AITasker.Domain.Entities;
using AITasker.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace AITasker.Infrastructure.Services;

public class AdminJobService : IAdminJobService
{
    private const string StatusDraft = "DRAFT";
    private const string StatusOpen = "OPEN";
    private const string StatusActive = "ACTIVE";
    private const string StatusCompleted = "COMPLETED";
    private const string StatusDisputed = "DISPUTED";
    private const string StatusCancelled = "CANCELLED";

    private const string ProposalStatusDraft = "DRAFT";
    private const string ProposalStatusSubmitted = "SUBMITTED";
    private const string ProposalStatusAccepted = "ACCEPTED";
    private const string ProposalStatusRejected = "REJECTED";
    private const string ProposalStatusWithdrawn = "WITHDRAWN";

    private readonly AITaskerDbContext _context;
    private readonly INotificationService _notificationService;
    private readonly IAdminAuditLogService _adminAuditLogService;

    public AdminJobService(
        AITaskerDbContext context,
        INotificationService notificationService,
        IAdminAuditLogService adminAuditLogService)
    {
        _context = context;
        _notificationService = notificationService;
        _adminAuditLogService = adminAuditLogService;
    }

    public async Task<List<AdminJobResponse>> GetJobsAsync(
        string? search,
        string? status,
        int? clientProfileId)
    {
        var query = _context.JobPostings
            .AsNoTracking()
            .Include(x => x.ClientProfile!)
                .ThenInclude(x => x.User)
            .Include(x => x.JobSkills)
                .ThenInclude(x => x.Skill)
            .AsQueryable();

        if (!string.IsNullOrWhiteSpace(search))
        {
            var trimmedSearch = search.Trim();

            query = query.Where(x =>
                x.Title.Contains(trimmedSearch) ||
                x.Description.Contains(trimmedSearch) ||
                (
                    x.ClientProfile != null &&
                    x.ClientProfile.User != null &&
                    (
                        x.ClientProfile.User.FullName.Contains(trimmedSearch) ||
                        x.ClientProfile.User.Email.Contains(trimmedSearch)
                    )
                ));
        }

        if (!string.IsNullOrWhiteSpace(status))
        {
            var normalizedStatus = status.Trim().ToUpperInvariant();
            query = query.Where(x => x.Status == normalizedStatus);
        }

        if (clientProfileId.HasValue)
        {
            query = query.Where(x => x.ClientProfileId == clientProfileId.Value);
        }

        var jobs = await query
            .OrderByDescending(x => x.CreatedAt)
            .ToListAsync();

        var proposalStats = await LoadProposalStatsAsync(
            jobs.Select(x => x.JobPostingId).ToList());

        return jobs
            .Select(job => MapJobToResponse(
                job,
                proposalStats.TryGetValue(job.JobPostingId, out var stats)
                    ? stats
                    : null))
            .ToList();
    }

    public async Task<AdminJobResponse?> GetJobByIdAsync(int jobPostingId)
    {
        var job = await _context.JobPostings
            .AsNoTracking()
            .Include(x => x.ClientProfile!)
                .ThenInclude(x => x.User)
            .Include(x => x.JobSkills)
                .ThenInclude(x => x.Skill)
            .FirstOrDefaultAsync(x => x.JobPostingId == jobPostingId);

        if (job == null)
        {
            return null;
        }

        var proposalStats = await LoadProposalStatsAsync(
            new List<int> { job.JobPostingId });

        return MapJobToResponse(
            job,
            proposalStats.TryGetValue(job.JobPostingId, out var stats)
                ? stats
                : null);
    }

    public async Task<IReadOnlyList<ProposalResponse>?> GetJobProposalsAsync(int jobPostingId)
    {
        var jobExists = await _context.JobPostings
            .AsNoTracking()
            .AnyAsync(x => x.JobPostingId == jobPostingId);

        if (!jobExists)
        {
            return null;
        }

        var proposals = await _context.Proposals
            .AsNoTracking()
            .Where(x =>
                x.JobId == jobPostingId &&
                x.Status != ProposalStatusDraft)
            .OrderByDescending(x => x.CreatedAt)
            .ToListAsync();

        var responses = new List<ProposalResponse>();

        foreach (var proposal in proposals)
        {
            responses.Add(await MapProposalToResponseAsync(proposal));
        }

        return responses;
    }

    public async Task<AdminJobResponse?> CancelJobAsync(
        int adminId,
        int jobPostingId,
        AdminCancelJobRequest request)
    {
        var reason = NormalizeCancelReason(request.Reason);

        var job = await _context.JobPostings
            .Include(x => x.ClientProfile!)
                .ThenInclude(x => x.User)
            .Include(x => x.JobSkills)
                .ThenInclude(x => x.Skill)
            .FirstOrDefaultAsync(x => x.JobPostingId == jobPostingId);

        if (job == null)
        {
            return null;
        }

        EnsureJobCanBeCancelledByAdmin(job);

        await using var transaction = await _context.Database.BeginTransactionAsync();

        var oldStatus = job.Status;
        var now = DateTime.UtcNow;

        job.Status = StatusCancelled;
        job.UpdatedAt = now;

        var submittedProposals = await _context.Proposals
            .Include(x => x.ExpertProfile!)
                .ThenInclude(x => x.User)
            .Where(x =>
                x.JobId == jobPostingId &&
                x.Status == ProposalStatusSubmitted)
            .ToListAsync();

        foreach (var proposal in submittedProposals)
        {
            proposal.Status = ProposalStatusRejected;
        }

        await _context.SaveChangesAsync();

        await NotifyClientJobCancelledAsync(job, reason);
        await NotifyExpertsJobCancelledAsync(job, submittedProposals, reason);

        await _adminAuditLogService.LogAsync(
            adminId,
            "CANCEL_JOB",
            "JobPosting",
            job.JobPostingId,
            oldStatus,
            StatusCancelled,
            reason);

        await transaction.CommitAsync();

        return await GetJobByIdAsync(job.JobPostingId)
            ?? throw new InvalidOperationException("Failed to load cancelled job.");
    }

    private static string NormalizeCancelReason(string? reason)
    {
        if (string.IsNullOrWhiteSpace(reason))
        {
            throw new InvalidOperationException("Cancel reason is required.");
        }

        var trimmedReason = reason.Trim();

        if (trimmedReason.Length > 500)
        {
            throw new InvalidOperationException("Cancel reason must not exceed 500 characters.");
        }

        return trimmedReason;
    }

    private static void EnsureJobCanBeCancelledByAdmin(JobPosting job)
    {
        var normalizedStatus = job.Status.Trim().ToUpperInvariant();

        if (normalizedStatus == StatusCancelled)
        {
            throw new InvalidOperationException("Job is already cancelled.");
        }

        if (normalizedStatus is StatusActive or StatusCompleted or StatusDisputed)
        {
            throw new InvalidOperationException(
                "This job is already linked to an active workflow and cannot be cancelled by Admin Job API.");
        }

        if (normalizedStatus is not (StatusDraft or StatusOpen))
        {
            throw new InvalidOperationException("Only draft or open jobs can be cancelled by admin.");
        }
    }

    private async Task NotifyClientJobCancelledAsync(
        JobPosting job,
        string reason)
    {
        if (job.ClientProfile == null)
        {
            return;
        }

        await _notificationService.CreateNotificationAsync(
            job.ClientProfile.UserId,
            "Job cancelled by admin",
            $"Your job \"{job.Title}\" has been cancelled by admin. Reason: {reason}",
            NotificationTypes.JobCancelledByAdmin,
            relatedEntityType: "JOB",
            relatedEntityId: job.JobPostingId,
            relatedJobId: job.JobPostingId);
    }

    private async Task NotifyExpertsJobCancelledAsync(
        JobPosting job,
        List<Proposal> submittedProposals,
        string reason)
    {
        var expertUserIds = submittedProposals
            .Select(x => x.ExpertProfile?.UserId)
            .Where(x => x.HasValue)
            .Select(x => x!.Value)
            .Distinct()
            .ToList();

        foreach (var expertUserId in expertUserIds)
        {
            await _notificationService.CreateNotificationAsync(
                expertUserId,
                "Job cancelled by admin",
                $"The job \"{job.Title}\" that you submitted a proposal to has been cancelled by admin. Reason: {reason}",
                NotificationTypes.JobCancelledForExpert,
                relatedEntityType: "JOB",
                relatedEntityId: job.JobPostingId,
                relatedJobId: job.JobPostingId);
        }
    }

    private async Task<Dictionary<int, ProposalStats>> LoadProposalStatsAsync(
        List<int> jobPostingIds)
    {
        if (jobPostingIds.Count == 0)
        {
            return new Dictionary<int, ProposalStats>();
        }

        var stats = await _context.Proposals
            .AsNoTracking()
            .Where(x => jobPostingIds.Contains(x.JobId))
            .GroupBy(x => x.JobId)
            .Select(group => new ProposalStats
            {
                JobId = group.Key,
                ProposalCount = group.Count(x => x.Status != ProposalStatusDraft),
                SubmittedProposalCount = group.Count(x => x.Status == ProposalStatusSubmitted),
                AcceptedProposalCount = group.Count(x => x.Status == ProposalStatusAccepted),
                RejectedProposalCount = group.Count(x => x.Status == ProposalStatusRejected),
                WithdrawnProposalCount = group.Count(x => x.Status == ProposalStatusWithdrawn)
            })
            .ToListAsync();

        return stats.ToDictionary(x => x.JobId);
    }

    private static AdminJobResponse MapJobToResponse(
        JobPosting job,
        ProposalStats? proposalStats)
    {
        var clientProfile = job.ClientProfile;
        var clientUser = clientProfile?.User;

        return new AdminJobResponse
        {
            JobPostingId = job.JobPostingId,
            ClientProfileId = job.ClientProfileId,
            ClientUserId = clientProfile?.UserId ?? 0,
            ClientName = clientUser?.FullName ?? string.Empty,
            ClientEmail = clientUser?.Email ?? string.Empty,
            ClientType = clientProfile?.ClientType ?? string.Empty,

            Title = job.Title,
            Description = job.Description,
            AiGeneratedDescription = job.AiGeneratedDescription,
            BudgetMin = job.BudgetMin,
            BudgetMax = job.BudgetMax,
            Deadline = job.Deadline,
            ProjectType = job.ProjectType,
            Complexity = job.Complexity,
            ExpectedDeliverables = job.ExpectedDeliverables,
            Status = job.Status,
            IsAiAssisted = job.IsAiAssisted,
            PostingChargeType = job.PostingChargeType,
            PublishedAt = job.PublishedAt,
            CreatedAt = job.CreatedAt,
            UpdatedAt = job.UpdatedAt,

            ProposalCount = proposalStats?.ProposalCount ?? 0,
            SubmittedProposalCount = proposalStats?.SubmittedProposalCount ?? 0,
            AcceptedProposalCount = proposalStats?.AcceptedProposalCount ?? 0,
            RejectedProposalCount = proposalStats?.RejectedProposalCount ?? 0,
            WithdrawnProposalCount = proposalStats?.WithdrawnProposalCount ?? 0,

            Skills = job.JobSkills
                .OrderBy(x => x.Skill != null ? x.Skill.SkillName : string.Empty)
                .Select(x => new JobSkillResponse
                {
                    SkillId = x.SkillId,
                    SkillName = x.Skill?.SkillName ?? string.Empty,
                    Category = x.Skill?.Category
                })
                .ToList()
        };
    }

    private async Task<ProposalResponse> MapProposalToResponseAsync(Proposal proposal)
    {
        var job = await _context.JobPostings
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.JobPostingId == proposal.JobId);

        if (job == null)
        {
            throw new InvalidOperationException("Job posting not found.");
        }

        var clientProfile = await _context.ClientProfiles
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.ClientProfileId == job.ClientProfileId);

        if (clientProfile == null)
        {
            throw new InvalidOperationException("Client profile not found.");
        }

        var clientUser = await _context.Users
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.UserId == clientProfile.UserId);

        if (clientUser == null)
        {
            throw new InvalidOperationException("Client user not found.");
        }

        var expertProfile = await _context.ExpertProfiles
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.ExpertProfileId == proposal.ExpertId);

        if (expertProfile == null)
        {
            throw new InvalidOperationException("Expert profile not found.");
        }

        var expertUser = await _context.Users
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.UserId == expertProfile.UserId);

        if (expertUser == null)
        {
            throw new InvalidOperationException("Expert user not found.");
        }

        var contractId = await _context.ProjectContracts
            .AsNoTracking()
            .Where(x => x.ProposalId == proposal.ProposalId)
            .Select(x => (int?)x.ContractId)
            .FirstOrDefaultAsync();

        var versions = await _context.ProposalVersions
            .AsNoTracking()
            .Where(x => x.ProposalId == proposal.ProposalId)
            .OrderByDescending(x => x.VersionNumber)
            .ToListAsync();

        var totalVersions = versions.Count == 0 ? 1 : versions.Count;
        var latestVersionNumber = versions.Count == 0 ? 1 : versions[0].VersionNumber;
        var latestVersion = versions.Count == 0
            ? null
            : await MapProposalVersionResponseAsync(versions[0]);

        var lastResubmittedAt = versions
            .Where(x => x.VersionNumber > 1)
            .OrderByDescending(x => x.CreatedAt)
            .Select(x => (DateTime?)x.CreatedAt)
            .FirstOrDefault();

        var milestoneDrafts = await _context.ProposalMilestoneDrafts
            .AsNoTracking()
            .Where(x => x.ProposalId == proposal.ProposalId)
            .OrderBy(x => x.OrderIndex)
            .ToListAsync();

        return new ProposalResponse
        {
            ProposalId = proposal.ProposalId,
            JobId = proposal.JobId,
            JobTitle = job.Title,

            ClientProfileId = clientProfile.ClientProfileId,
            ClientUserId = clientProfile.UserId,
            ClientName = clientUser.FullName,

            ExpertProfileId = expertProfile.ExpertProfileId,
            ExpertUserId = expertProfile.UserId,
            ExpertName = expertUser.FullName,

            CoverLetter = proposal.CoverLetter,
            ProposedPrice = proposal.ProposedPrice,
            ProposedTimelineDays = proposal.ProposedTimelineDays,
            ExpectedOutputs = proposal.ExpectedOutputs,
            WorkingApproach = proposal.WorkingApproach,
            PreliminaryMilestonePlan = proposal.PreliminaryMilestonePlan,
            Milestones = MapProposalMilestoneDraftResponses(milestoneDrafts),

            Status = proposal.Status,
            ContractId = contractId,

            LatestVersionNumber = latestVersionNumber,
            TotalVersions = totalVersions,
            LastResubmittedAt = lastResubmittedAt,
            ResubmitLimit = 0,
            ResubmitWindowHours = 0,
            RemainingResubmitsInWindow = int.MaxValue,
            LatestVersion = latestVersion,

            CreatedAt = proposal.CreatedAt
        };
    }

    private async Task<ProposalVersionResponse> MapProposalVersionResponseAsync(
        ProposalVersion version)
    {
        var createdBy = await _context.Users
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.UserId == version.CreatedByUserId);

        return new ProposalVersionResponse
        {
            ProposalVersionId = version.ProposalVersionId,
            ProposalId = version.ProposalId,
            VersionNumber = version.VersionNumber,
            CoverLetter = version.CoverLetter,
            ProposedPrice = version.ProposedPrice,
            ProposedTimelineDays = version.ProposedTimelineDays,
            ExpectedOutputs = version.ExpectedOutputs,
            WorkingApproach = version.WorkingApproach,
            PreliminaryMilestonePlan = version.PreliminaryMilestonePlan,
            MilestonePlanJson = version.MilestonePlanJson,
            ResubmitNote = version.ResubmitNote,
            CreatedByUserId = version.CreatedByUserId,
            CreatedByName = createdBy?.FullName ?? string.Empty,
            CreatedAt = version.CreatedAt
        };
    }

    private static List<ProposalMilestoneDraftResponse> MapProposalMilestoneDraftResponses(
        List<ProposalMilestoneDraft> drafts)
    {
        var result = new List<ProposalMilestoneDraftResponse>();
        var previousDeadlineOffsetDays = 0;

        foreach (var draft in drafts.OrderBy(x => x.OrderIndex))
        {
            var durationDays = draft.DeadlineOffsetDays - previousDeadlineOffsetDays;

            result.Add(new ProposalMilestoneDraftResponse
            {
                ProposalMilestoneDraftId = draft.ProposalMilestoneDraftId,
                ProposalId = draft.ProposalId,
                Title = draft.Title,
                Amount = draft.Amount,
                DurationDays = durationDays <= 0 ? draft.DeadlineOffsetDays : durationDays,
                CreatedAt = draft.CreatedAt
            });

            previousDeadlineOffsetDays = draft.DeadlineOffsetDays;
        }

        return result;
    }

    private sealed class ProposalStats
    {
        public int JobId { get; set; }

        public int ProposalCount { get; set; }

        public int SubmittedProposalCount { get; set; }

        public int AcceptedProposalCount { get; set; }

        public int RejectedProposalCount { get; set; }

        public int WithdrawnProposalCount { get; set; }
    }
}
